<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <meta name="description" content="Documentation for the ProHap tool">
    <meta name="author" content="Jakub Vasicek">
    <!-- Bootstrap core CSS -->
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@5.2.3/dist/css/bootstrap.min.css" integrity="sha384-rbsA2VBKQhggwzxH7pPCaAqO46MgnOM80zW1RWuH61DGLwZJEdK2Kadq2F9CUG65" crossorigin="anonymous">
</head>

<body class="bg-light">
    <div class="container">
    <div class="py-5 text-left">
        <h2>Create your config.yaml file for <a href="https://github.com/ProGenNo/ProHap">ProHap</a></h2>        
    </div>
    <div class="row">
        <div class="col">
            <form id="form-general-params">
                <h3 class="mb-4">General parameters</h3>                
                <hr class="mb-1">
                <div class="row">
                    <div class="col-md-4">
                        <label for="ensembl-release">Ensembl release</label>
                        <input type="text" class="form-control very-short-input" name="ensembl-release" value="113" id="ensembl-release" required="required" aria-required="true">
                    </div>
                </div>
                <div class="row">
                    <div class="col-md-8">
                        <hr class="mb-1">
                        <legend class="col-form-label col-sm-2 pt-0">Select transcripts</legend>
                        <div class="d-block my-3">
                            <div class="form-check">
                                <input class="form-check-input" type="radio" name="transcript-radio" id="transcript-radio-1" value="default" checked>
                                <label class="form-check-label" for="transcript-radio-1">Use the default set of transcripts</label>
                            </div>
                            <div class="form-check">
                                <input class="form-check-input" type="radio" name="transcript-radio" id="transcript-radio-2" value="default">
                                <label class="form-check-label" for="transcript-radio-2">Use only <a href="https://www.ncbi.nlm.nih.gov/refseq/MANE/">MANE Select</a> transcripts</label>
                                <br>
                                <span class="text-muted">Only available for Ensembl v.108 and above. For genes that do not include any MANE Select transcript in Ensembl, "Ensembl Canonical" transcripts will be selected.</span>
                            </div>
                            <div class="form-check">
                                <input class="form-check-input" type="radio" name="transcript-radio" id="transcript-radio-3" value="biotype">
                                <label class="form-check-label" for="transcript-radio-3">Select transcripts by biotype (provide below)</label>
                            </div>
                            <div class="form-check">
                                <input class="form-check-input" type="radio" name="transcript-radio" id="transcript-radio-4" value="custom">
                                <label class="form-check-label" for="transcript-radio-4">User-defined list of transcripts (provide below)</label>
                            </div>
                        </div>
                    </div>
                </div>
                <div class="row">
                    <div class="col-md-8">
                        <div class="d-block my-1">
                            <label for="transcript-biptypes">
                                Transcript biotypes to be included
                                <span class="text-muted">(Comma-separated list, use biotypes from <a href="https://www.gencodegenes.org/pages/biotypes.html">Gencode</a>)</span>
                            </label>
                            <input type="text" class="form-control" name="transcript-biotypes" placeholder="e.g. protein_coding,retained_intron" id="transcript-biotypes" disabled>
                        </div>
                    </div>
                </div>
                <div class="row">
                    <div class="col-md-8">
                        <div class="d-block my-1">
                            <label for="transcript-file-path">
                                Path to the custom transcript list
                                <span class="text-muted">(CSV file, see example <a href="https://github.com/ProGenNo/ProHap/blob/main/data/transcripts_reference_108.csv">data/transcripts_reference_108.csv</a>)</span>
                            </label>
                            <input type="text" class="form-control" name="transcript-file-path" value="" id="transcript-file-path" disabled>
                        </div>
                        <hr class="mb-1">
                    </div>
                </div>
                <div class="row">
                    <div class="col-md-8">
                        <div class="d-block my-1">
                            <label for="contam-fasta-path">
                                Path to the contaminant FASTA file
                            </label>
                            <input type="text" class="form-control" name="contam-fasta-path" value="crap.fasta" id="contam-fasta-path" required="required" aria-required="true">
                            <span class="text-muted">The <a href="https://www.thegpm.org/crap/" target="_blank">default contaminant database</a> is provided in the crap.fasta file in this repository.</span>
                        </div>
                    </div>
                </div>
                <div class="row">
                    <div class="col-md-8">
                        <div class="d-block my-1">
                            <label for="final-fasta-path">
                                Path to the final FASTA file
                            </label>
                            <input type="text" class="form-control" name="final-fasta-path" value="" id="final-fasta-path" required="required" aria-required="true" placeholder="e.g. results/final_database.fa">
                        </div>
                    </div>
                </div>
                <div class="row">
                    <div class="form-check form-switch">
                        <input class="form-check-input" type="checkbox" id="simplify-switch">
                        <label class="form-check-label" for="simplify-switch">Simplify FASTA headers <span class="text-muted">(extract all information from the FASTA protein headers in to a tab-separated file)</span></label>
                    </div>
                </div>
                <hr class="mb-1">
            </form>
        </div>
    </div>
    <div class="row">
        <div class="col col-md-6">
            <form id="form-prohap-params">
                <h3 class="mb-4 text-center">ProHap</h3>  
                <hr class="mb-3">
                <div class="form-check form-switch">
                    <input class="form-check-input" type="checkbox" id="prohap-switch" checked>
                    <label class="form-check-label" for="prohap-switch">Use ProHap <span class="text-muted">(Include protein haplotypes in the final FASTA file)</span></label>
                </div>
                <hr class="mb-1">
                <legend class="col-form-label col-md-2 pt-0">Data source:</legend>
                <div class="d-block my-3">
                    <div class="form-check">
                        <input class="form-check-input" type="radio" name="prohap-source-radio" id="prohap-source-radio-1" value="url" checked>
                        <label class="form-check-label" for="prohap-source-radio-1">Download VCF files from an online resource</label>
                    </div>
                    <div class="form-check">
                        <input class="form-check-input" type="radio" name="prohap-source-radio" id="prohap-source-radio-2" value="local">
                        <label class="form-check-label" for="prohap-source-radio-2">Provide VCF files locally</label>
                    </div>
                </div>
                <hr class="mb-1">
                <label for="phased-vcf-url">URL of the data set of phased genotpyes</label>
                <input type="text" class="form-control" name="phased-vcf-url" value="http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/1000_genomes_project/release/20190312_biallelic_SNV_and_INDEL/" id="phased-vcf-url" required="required" aria-required="true">
                <small class="form-text text-muted">Default: 1000 Genomes Project on GRCh38</small>
                <div class="mb-1"></div>
                <label for="phased-vcf-url">Path to the directory containing phased VCF files</label>
                <input type="text" class="form-control" name="phased-vcf-path" placeholder="e.g. /mnt/c/data/my_VCF_dataset/" value="" id="phased-vcf-path" required="required" aria-required="true" disabled>
                <div class="mb-1"></div>
                <label for="phased-vcf-filename">Name of the VCF files
                </label>
                <input type="text" class="form-control" name="phased-vcf-filename" value="ALL.chr{chr}.shapeit2_integrated_snvindels_v2a_27022019.GRCh38.phased.vcf.gz" id="phased-vcf-filename" required="required" aria-required="true">
                <small class="form-text text-muted">VCFs are expected per chromosome, replace the chromosome number with "{chr}". Files can be either in the GZIP format or uncompressed.</small>
                <div class="mb-1"></div>
                <label for="sample_metadata_file">Samples metadata file
                </label>
                <input type="text" class="form-control" name="sample_metadata_file" value="igsr_samples.tsv" id="sample_metadata_file" required="required" aria-required="true">
                <small class="form-text text-muted">See the <a href="https://github.com/ProGenNo/ProHap/wiki/Input-&-Usage#prohap">wiki page</a> for details</small>
                
                <hr class="mb-1">
                <label for="phased-min-af">
                    MAF threshold
                </label>
                <input type="text" class="form-control very-short-input" name="phased-min-af" value="0.01" id="phased-min-af" required="required" aria-required="true">
                <small class="form-text text-muted">Variants under this threshold will not be included in haplotypes</small>
                <hr class="mb-1">
                <label for="phased-af-field">
                    MAF field name
                </label>
                <input type="text" class="form-control very-short-input" name="phased-af-field" value="AF" id="phased-af-field" required="required" aria-required="true">
                <small class="form-text text-muted">Name of the AF column in the VCF file ("AF" by default). Change if you want to use the frequency in a specific population within 1000 Genomes, or according to your own file </small>
                <hr class="mb-1">
                <legend class="col-form-label col-md-4 pt-0">Threshold haplotypes by</legend>
                <div class="d-block my-1">
                    <div class="form-check form-check-inline">
                        <input class="form-check-input" type="radio" name="haplo-threshold-by" id="haplo-threshold-by-1" value="default" checked>
                        <label class="form-check-label" for="haplo-threshold-by-1">Haplotype frequency</label>
                    </div>
                    <div class="form-check form-check-inline">
                        <input class="form-check-input" type="radio" name="haplo-threshold-by" id="haplo-threshold-by-2" value="biotype">
                        <label class="form-check-label" for="haplo-threshold-by-2">Haplotype occurrence count</label>
                    </div>
                </div>
                <label for="haplo-threshold">
                    Threshold value
                </label>
                <input type="text" class="form-control very-short-input" name="haplo-threshold" value="0" id="haplo-threshold" required="required" aria-required="true">
                <small class="form-text text-muted">Specify 0 to skip haplotype thresholding</small>
                <hr class="mb-1">
                <div class="row">
                    <legend class="col-form-label pt-0">Pseudo-autosomal regions (PAR) on the X chromosome</legend>
                </div>
                <div class="row">
                    <div class="col-md-4">
                        <label for="x_par1_to">
                            End of PAR1:
                        </label>
                        <input type="text" class="form-control" name="x_par1_to" value="2781479" id="x_par1_to" required="required" aria-required="true">
                    </div>
                    <div class="col-md-4">
                        <label for="x_par2_from">
                            Start of PAR2:
                        </label>
                        <input type="text" class="form-control" name="x_par2_from" value="155701383" id="x_par2_from" required="required" aria-required="true">
                    </div>
                </div>
                <small class="form-text text-muted">The default values for the GRCh38 human genome are 2781479 and 155701383. For GRCh37, use 2699520 and 154931044 respectively.</small>
                <hr class="mb-1">
                <div class="form-check form-switch">
                    <input class="form-check-input" type="checkbox" id="haplo_require_start" checked>
                    <label class="form-check-label" for="haplo_require_start">Require annotation of the start codon in transcripts</label><br>
                    <small class="form-text text-muted">Transcripts that do not have an annotated canonical start codon will not be used.</small>
                </div>
                <div class="form-check form-switch">
                    <input class="form-check-input" type="checkbox" id="haplo_ignore_UTR" checked>
                    <label class="form-check-label" for="haplo_ignore_UTR">Ignore variation in UTR regions</label><br>
                    <small class="form-text text-muted">If disabled, UTR sequences are still removed in the final optimized database, but retained in the haplotypes FASTA.</small>
                </div>
                <div class="form-check form-switch">
                    <input class="form-check-input" type="checkbox" id="haplo_skip_start_lost" checked>
                    <label class="form-check-label" for="haplo_skip_start_lost">Skip haplotypes where the start codon is lost</label><br>
                    <small class="form-text text-muted">If disabled, these haplotype cDNA sequences are translated in 3 reading frames, including UTR sequences.</small>
                </div>
                <div class="form-check form-switch">
                    <input class="form-check-input" type="checkbox" id="haplo_output_cdnas">
                    <label class="form-check-label" for="haplo_output_cdnas">Output haplotype cDNA sequences</label><br>
                    <small class="form-text text-muted">Create a separate file containing all the haplotype cDNA sequences before translation. If skipping UTR variation as above, the cDNA haplotypes will begin with the canonical start codon.</small>
                </div>
                <hr class="mb-1">
                
                <label for="haplo_cdna_file">
                    Path to the cDNA haplotype FASTA file
                </label>
                <input type="text" class="form-control" name="haplo_cdna_file" value="" id="haplo_cdna_file" required="required" aria-required="true" placeholder="e.g. results/haplotypes_cdna.fa" disabled>
                
                <label for="haplo_fasta_file">
                    Path to the protein haplotype FASTA file
                </label>
                <input type="text" class="form-control" name="haplo_fasta_file" value="" id="haplo_fasta_file" required="required" aria-required="true" placeholder="e.g. results/haplotypes.fa">
                
                <div class="mb-1"></div>
                <label for="haplo_table_file">
                    Path to the haplotype metadata table
                </label>
                <input type="text" class="form-control" name="haplo_table_file" value="" id="haplo_table_file" required="required" aria-required="true" placeholder="e.g. results/haplotypes.tsv">
            </form>            
        </div>
        <div class="col col-md-6">
            <form id="form-provar-params">
                <h3 class="mb-4 text-center">ProVar</h3>
                <hr class="mb-3">
                <div class="form-check form-switch">
                    <input class="form-check-input" type="checkbox" id="provar-switch">
                    <label class="form-check-label" for="provar-switch">Use ProVar <span class="text-muted">(Include individual variants in the final FASTA file)</span></label>
                </div>
                <hr class="mb-1">
                <h5 class="mb-3 text-left">Add your VCF files:</h5>  
                <div id="vcf-file-specs">  
                    <div class="row">
                        <div class="col-md-3">
                            <label for="VCF-name-1">Dataset name</label>
                            <input type="text" class="form-control" name="VCF-name-1" value="" id="VCF-name-1" required="required" aria-required="true" placeholder="e.g. RareVariants">
                        </div>
                        <div class="col-md-4">
                            <label for="VCF-path-1">VCF file path</label>
                            <input type="text" class="form-control" name="VCF-path-1" value="" id="VCF-path-1" required="required" aria-required="true" placeholder="e.g. data/inhouse.vcf">
                        </div>
                        <div class="col-md-4">
                            <label for="VCF-MAF-1">
                                MAF threshold
                            </label>
                            <input type="text" class="form-control short-input" name="VCF-MAF-1" value="0" id="VCF-MAF-1" required="required" aria-required="true">
                            <small class="form-text text-muted">Specify 0 to skip thresholding</small>
                        </div>
                    </div>
                </div>
                <div class="mb-1"></div>
                <button type="button" class="btn btn-outline-primary" id="add-vcf-file">Add</button>
                <hr class="mb-1">
                <div class="form-check form-switch">
                    <input class="form-check-input" type="checkbox" id="var_require_start" checked>
                    <label class="form-check-label" for="var_require_start">Require annotation of the start codon in transcripts</label><br>
                    <small class="form-text text-muted">Transcripts that do not have an annotated canonical start codon will not be used.</small>
                </div>                
                <div class="form-check form-switch">
                    <input class="form-check-input" type="checkbox" id="var_output_cdnas">
                    <label class="form-check-label" for="var_output_cdnas">Output variant cDNA sequences</label><br>
                    <small class="form-text text-muted">Create a separate file containing all the variant cDNA sequences before translation.</small>
                </div>
                <hr class="mb-1">
                <label for="var_cdna_file">
                    Path to the variant cDNA FASTA file
                </label>
                <input type="text" class="form-control" name="var_cdna_file" value="" id="var_cdna_file" required="required" aria-required="true" placeholder="e.g. results/variants_cdna.fa" disabled>
              
                <div class="mb-1"></div>
                <label for="var_fasta_file">
                    Path to the variant FASTA file
                </label>
                <input type="text" class="form-control" name="var_fasta_file" value="" id="var_fasta_file" required="required" aria-required="true" placeholder="e.g. results/variants.fa">
                
                <div class="mb-1"></div>
                <label for="var_table_file">
                    Path to the variant metadata table
                </label>
                <input type="text" class="form-control" name="var_table_file" value="" id="var_table_file" required="required" aria-required="true" placeholder="e.g. results/variants.tsv">
                
                <hr class="mb-1">
                <div class="form-check form-switch">
                    <input class="form-check-input" type="checkbox" id="add_existing_haplo">
                    <label class="form-check-label" for="add_existing_haplo">Merge with an existing protein haplotype database</label>
                </div>
                <div class="mb-1"></div>
                <label for="haplo_added_table">
                    Path to the additional haplotype table file
                </label>
                <input type="text" class="form-control" name="haplo_added_table" value="" id="haplo_added_table" aria-required="false" placeholder="e.g. data/F2_231220_1kGP_GRCh38_haplo.tsv">
                <small class="form-text text-muted">(e.g., one of the F2 files in the <a href="https://doi.org/10.5281/zenodo.10149278">Zenodo repository</a>)</small>
                <div class="mb-1"></div>
                <label for="haplo_added_fasta">
                    Path to the additional haplotype FASTA file
                </label>
                <input type="text" class="form-control" name="haplo_added_fasta" value="" id="haplo_added_fasta" aria-required="false" placeholder="e.g. data/F3_231220_1kGP_GRCh38_haplo.fa">
                <small class="form-text text-muted">(e.g., one of the F3 files in the <a href="https://doi.org/10.5281/zenodo.10149278">Zenodo repository</a>)</small>
            </form>
        </div>
    </div>
    <hr class="mb-3">
    <div class="mb-3">
        <button onclick="download('config.yaml')">Download</button>
        <span>or copy the content below to your config.yaml file:</span>
    </div>
    <div class="row">
        <div class="col" id="config-content">            
        </div>
    </div>
</div>

<footer class="my-5 pt-5 text-muted text-center">
    <p class="mb-1">2025 University of Bergen</p>
</footer>

<style>
    .container {
        margin-top: 5vh;
    }
    .short-input {
        width: 35%;
    }
    .very-short-input {
        width: 15%;
    }
    #config-content {
        background: #bbb;
    }
</style>

<script src="https://d3js.org/d3.v7.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/bootstrap@5.2.3/dist/js/bootstrap.min.js" integrity="sha384-cuYeSxntonz0PPNlHhBs68uyIAVpIIOZZ5JqeqvYYIcEL727kskC66kF92t6Xl2V" crossorigin="anonymous"></script>
<script src="configBuilder.js"></script>

<script lang="ts">
    let general_config_html = setGeneralParams("112", "data/transcripts_reference_112.csv", false, "", "crap.fasta", "", false)
    let provar_config_html = setProVarParams(false, [], true, false, '', '', '', false, '', '')
    let prohap_config_html = setProHapParams( 
        true, 
        "http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/1000_genomes_project/release/20190312_biallelic_SNV_and_INDEL/", 
        "",
        "ALL.chr{chr}.shapeit2_integrated_snvindels_v2a_27022019.GRCh38.phased.vcf", "igsr_samples.tsv",
        "-1", "0", "0.01", "AF", "2781479", "155701383", true, true, true, false, "", "", ""
    )

    function update_config_text() {
        d3.select('#config-content').html(general_config_html + prohap_config_html + provar_config_html)
    }

    function download(filename) {
        text = (general_config_html + prohap_config_html + provar_config_html).replaceAll('<p>', '').replaceAll('</p>', '\n').replaceAll('<br>', '\n')

        var element = document.createElement('a');
        element.setAttribute('href', 'data:text/plain;charset=utf-8,' + encodeURIComponent(text));
        element.setAttribute('download', filename);

        element.style.display = 'none';
        document.body.appendChild(element);

        element.click();

        document.body.removeChild(element);
    }

    update_config_text()

    d3.select('#form-general-params').on('change', (evt) => {
        const ensembl_release = evt.currentTarget[0].value
        let only_MANE = false

        let custom_tr_list = ""
        let transcript_biotypes = ""
        if (evt.currentTarget[1].checked) {
            custom_tr_list = "data/transcripts_reference_" + ensembl_release + ".csv"
            d3.select('#transcript-biotypes').attr('disabled', true)
            d3.select('#transcript-file-path').attr('disabled', true)
        } else if (evt.currentTarget[2].checked) {
            custom_tr_list = "data/transcripts_reference_" + ensembl_release + ".csv"
            only_MANE = true
            d3.select('#transcript-biotypes').attr('disabled', true)
            d3.select('#transcript-file-path').attr('disabled', true)
        } else if (evt.currentTarget[3].checked) {
            transcript_biotypes = evt.currentTarget[5].value
            d3.select('#transcript-biotypes').attr('disabled', null)
            d3.select('#transcript-file-path').attr('disabled', true)
        } else if (evt.currentTarget[4].checked) {
            custom_tr_list = evt.currentTarget[6].value
            d3.select('#transcript-biotypes').attr('disabled', true)
            d3.select('#transcript-file-path').attr('disabled', null)
        }

        const contam_fasta_path = evt.currentTarget[7].value
        const final_fasta_path = evt.currentTarget[8].value
        const simplify_headers = evt.currentTarget[9].checked

        general_config_html = setGeneralParams(ensembl_release, custom_tr_list, only_MANE, transcript_biotypes, contam_fasta_path, final_fasta_path, simplify_headers)
        update_config_text()
    })

    d3.select('#form-provar-params').on('change', (evt) => {

        const include_variants = evt.currentTarget[0].checked

        let VCF_data = []

        let VCFname = ""
        let VCFpath = ""

        let vcf_form_elements = 0   // counter for the VCF attribute elements in the form

        for (let element_idx in evt.currentTarget) {
            const element = evt.currentTarget[parseInt(element_idx)]

            if (element && element.id.startsWith('VCF')) {
                vcf_form_elements += 1

                if (element.id.includes('name')) {
                    VCFname = element.value
                } else if (element.id.includes('path')) {
                    VCFpath = element.value
                } else if (element.id.includes('MAF')) {
                    if ((VCFname !== "") && (VCFpath !== "")) {
                        VCF_data.push([VCFname, VCFpath, element.value.length === 0 ? "0" : element.value])
                    }

                    VCFname = ""
                    VCFpath = ""
                }
            }
        }

        const var_require_start = evt.currentTarget[2 + vcf_form_elements].checked        
        const var_output_cdna = evt.currentTarget[3 + vcf_form_elements].checked

        let var_cdna_file = ""

        if (var_output_cdna) {
            d3.select('#var_cdna_file').attr('disabled', null)
            var_cdna_file = evt.currentTarget[4 + vcf_form_elements].value
        } else {
            d3.select('#var_cdna_file').attr('disabled', true)
        }

        const var_fasta_file = evt.currentTarget[5 + vcf_form_elements].value
        const var_table_file = evt.currentTarget[6 + vcf_form_elements].value

        const add_existing_haplo = evt.currentTarget[7 + vcf_form_elements].checked
        const haplo_added_table = evt.currentTarget[8 + vcf_form_elements].value
        const haplo_added_fasta = evt.currentTarget[9 + vcf_form_elements].value

        provar_config_html = setProVarParams(include_variants, VCF_data, var_require_start, var_output_cdna, var_cdna_file, var_fasta_file, var_table_file, add_existing_haplo, haplo_added_table, haplo_added_fasta)
        update_config_text()
    })

    d3.select('#form-prohap-params').on('change', (evt) => {
        const include_haplotypes = evt.currentTarget[0].checked
        
        let dataset_url = ""
        let dataset_path = ""
        
        if (evt.currentTarget[1].checked) {
            d3.select('#phased-vcf-url').attr('disabled', null)
            d3.select('#phased-vcf-path').attr('disabled', true)
            dataset_url = evt.currentTarget[3].value
        } else if (evt.currentTarget[2].checked) {
            d3.select('#phased-vcf-url').attr('disabled', true)
            d3.select('#phased-vcf-path').attr('disabled', null)
            dataset_path = evt.currentTarget[4].value
        } 

        const vcf_filename = evt.currentTarget[5].value
        const samples_filename = evt.currentTarget[6].value

        const MAF_threshold = evt.currentTarget[7].value
        const MAF_field = evt.currentTarget[8].value

        let freq_threshold = "-1"
        let count_threshold = "0"

        if (evt.currentTarget[9].checked) {
            freq_threshold = evt.currentTarget[11].value
        } else {
            count_threshold = evt.currentTarget[11].value
        }

        const x_par1_to = evt.currentTarget[12].value
        const x_par2_from = evt.currentTarget[13].value

        const haplo_require_start = evt.currentTarget[14].checked
        const haplo_ignore_UTR = evt.currentTarget[15].checked
        const haplo_skip_start_lost = evt.currentTarget[16].checked
        const haplo_output_cdnas = evt.currentTarget[17].checked

        let haplo_cdna_file = ""

        if (haplo_output_cdnas) {
            d3.select('#haplo_cdna_file').attr('disabled', null)
            haplo_cdna_file = evt.currentTarget[18].value
        } else {
            d3.select('#haplo_cdna_file').attr('disabled', true)
        }

        const haplo_fasta_file = evt.currentTarget[19].value
        const haplo_table_file = evt.currentTarget[20].value

        prohap_config_html = setProHapParams(include_haplotypes, dataset_url, dataset_path, vcf_filename, samples_filename, freq_threshold, count_threshold, MAF_threshold, MAF_field, x_par1_to, x_par2_from, haplo_require_start, haplo_ignore_UTR, haplo_skip_start_lost, haplo_output_cdnas, haplo_cdna_file, haplo_fasta_file, haplo_table_file)
        update_config_text()
    })

    let variant_vcf_input_count = 1

    d3.select('#add-vcf-file').on('click', (evt) => {
        variant_vcf_input_count += 1
        d3.select('#vcf-file-specs').html(
            d3.select('#vcf-file-specs').html() + 
            "<div class=\"mb-1\"></div>" +
            "\
            <div class=\"row\" id=\"vcf-input-row-"+ variant_vcf_input_count +"\">\
                <div class=\"col-md-3\">\
                    <input type=\"text\" class=\"form-control\" name=\"VCF-name-"+ variant_vcf_input_count +"\" value=\"\" id=\"VCF-name-"+ variant_vcf_input_count +"\" required=\"required\" aria-required=\"true\" placeholder=\"e.g. RareVariants\">\
                </div>\
                <div class=\"col-md-4\">\
                    <input type=\"text\" class=\"form-control\" name=\"VCF-path-"+ variant_vcf_input_count +"\" value=\"\" id=\"VCF-path-"+ variant_vcf_input_count +"\" required=\"required\" aria-required=\"true\" placeholder=\"e.g. data/inhouse.vcf\">\
                </div>\
                <div class=\"col-md-4\">\
                    <input type=\"text\" class=\"form-control short-input\" name=\"VCF-MAF-"+ variant_vcf_input_count +"\" value=\"0\" id=\"VCF-MAF-"+ variant_vcf_input_count +"\" required=\"required\" aria-required=\"true\">\
                </div>\
                <div class=\"col-sm-1\">\
                    <button type=\"button\" class=\"btn btn-outline-danger\" id=\"VCF-row-rm-"+ variant_vcf_input_count +"\">X</button>\
                </div>\
            </div>\
            "
        )

        for (let rownr=2; rownr <= variant_vcf_input_count; rownr++) {
            d3.select("#VCF-row-rm-"+ rownr).on('click', evt => {
                d3.select("#vcf-input-row-"+ rownr).remove()
                d3.select('#form-provar-params').change()
            })
        }
    })
</script>
</body>
</html>